### PR TITLE
Feature/non media sync

### PR DESCRIPTION
**Pull request by DavidCramer**

### CODE REVIEW

**Review by pereirinha — COMMENTED** [review]
@DavidCramer 

Great stuff.
I left you some comments.

I'll be starting to test the functionality.

File: php/assets/class-rest-assets.php [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	 * Register the endpoints.
+	 *
+	 * @param array $endpoints The endpoint to add to.
+	 *
+	 * @return array
+	 */
+	public function rest_endpoints( $endpoints ) {
+
+		$endpoints['show_cache']          = array(
+			'method'              => \WP_REST_Server::CREATABLE,
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/assets/class-rest-assets.php:67]: @pereirinha - technically yes, but the fetch is using parameters as a post.

File: php/assets/class-rest-assets.php [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+	public function rest_purge( $request ) {
+
+		$asset      = $request->get_param( 'cachePoint' );
+		$query_args = array(
+			'post_type'      => Assets::POST_TYPE_SLUG,
+			'posts_per_page' => 20,
+			'paged'          => 1,
+			'post_parent'    => $asset,
+			'post_status'    => array( 'inherit', 'draft' ),
+			'fields'         => 'ids',
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/assets/class-rest-assets.php:115]: @pereirinha addressed this in PR https://github.com/cloudinary/cloudinary_wordpress/pull/490

File: php/assets/class-rest-assets.php [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+	/**
+	 * Admin permission callback.
+	 *
+	 * Explicitly defined to allow easier testability.
+	 *
+	 * @return bool
+	 */
+	public function rest_can_manage_options() {
+		return current_user_can( 'manage_options' );
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/assets/class-rest-assets.php:159]: @pereirinha - I  agree, but not needed now. I would prefer to add that to the main rest component and set these from a method in there.

File: php/class-delivery.php [hunk 4]
```diff
@@ -94,6 +95,23 @@ protected function setup_hooks() {
...
 
+	/**
+	 * Add DNS prefetch link tag for assets.
+	 *
+	 * @param array  $urls          URLs to print for resource hints.
+	 * @param string $relation_type The relation type the URLs are printed for, e.g. 'preconnect' or 'prerender'.
+	 *
+	 * @return array
+	 */
+	public function dns_prefetch( $urls, $relation_type ) {
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/class-delivery.php:106]: @pereirinha - well, I had it in, but forgot to add it to the delivery 🤦  https://github.com/cloudinary/cloudinary_wordpress/blob/master/php/cache/class-cache-point.php#L124-L139

File: php/settings/class-setting.php [hunk 5]
```diff
@@ -330,6 +330,30 @@ public function get_settings() {
...
 	}
 
+	/**
+	 * Get settings recursively of a specific type.
+	 *
+	 * @param string $type The type of settings to get.
+	 *
+	 * @return array
+	 */
+	public function get_settings_by_type( $type ) {
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/settings/class-setting.php:340]: As described, it's a helper function to get by type.

File: php/class-assets.php [hunk 6]
```diff
@@ -0,0 +1,1248 @@
...
+		add_filter( 'cloudinary_is_local_asset_url', array( $this, 'check_asset' ), 10, 2 );
+		add_filter( 'cloudinary_delivery_get_id', array( $this, 'get_asset_id_from_tag' ), 10, 2 );
+		add_filter( 'cloudinary_is_media', array( $this, 'is_media' ), 10, 2 );
+		add_filter( 'get_attached_file', array( $this, 'get_attached_file' ), 10, 2 );
+		add_filter( 'cloudinary_sync_base_struct', array( $this, 'add_sync_type' ) );
+		add_filter( 'update_post_metadata', array( $this, 'update_meta' ), 10, 4 );
+		add_filter( 'add_post_metadata', array( $this, 'update_meta' ), 10, 4 );
+		add_filter( 'get_post_metadata', array( $this, 'get_meta' ), 10, 3 );
+		add_filter( 'delete_post_metadata', array( $this, 'delete_meta' ), 10, 4 );
+		add_filter( 'intermediate_image_sizes_advanced', array( $this, 'no_sizes' ), PHP_INT_MAX, 3 );
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/class-assets.php:148]: @pereirinha - I agree, let's make a ticket for it.

File: php/class-assets.php [hunk 7]
```diff
@@ -0,0 +1,1248 @@
...
+	 */
+	public function create_asset_parent( $path, $version ) {
+		$args      = array(
+			'post_title'  => trailingslashit( $path ),
+			'post_name'   => md5( $path ),
+			'post_type'   => self::POST_TYPE_SLUG,
+			'post_status' => 'publish',
+		);
+		$parent_id = wp_insert_post( $args );
+		if ( $parent_id ) {
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/class-assets.php:415]: @pereirinha only if you set the second argument to true (`$wp_error = false`)

File: php/class-assets.php [hunk 8]
```diff
@@ -0,0 +1,1248 @@
...
+	 * @param int $asset_id The asset ID to upload.
+	 *
+	 * @return array|\WP_Error
+	 */
+	public function upload( $asset_id ) {
+		$connect           = $this->plugin->get_component( 'connect' );
+		$asset             = get_post( $asset_id );
+		$path              = trim( wp_normalize_path( str_replace( home_url(), '', $asset->post_title ) ), '/' );
+		$info              = pathinfo( $path );
+		$cloudinary_folder = $this->media->get_cloudinary_folder( true );
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/class-assets.php:450]: my bad, I forgot

File: php/class-assets.php [hunk 9]
```diff
@@ -0,0 +1,1248 @@
...
+	 * @param string $filename The filename to check.
+	 *
+	 * @return bool
+	 */
+	protected function syncable_asset( $filename ) {
+		static $allowed_kinds = array();
+		if ( empty( $allowed_kinds ) ) {
+			// Check with paths.
+			$types           = wp_get_ext_types();
+			$allowed_kinds   = array_merge( $allowed_kinds, $types['image'] );
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/class-assets.php:554]: @pereirinha well, now that you mention it, yes 😂

**Review by pereirinha — APPROVED** [review]
@DavidCramer 

Thanks. This is looking great.
I'll do some testing and merge if all is good.

File: php/media/class-filter.php [hunk 10]
```diff
@@ -734,6 +734,23 @@ public function init_rest_filters() {
...
+	 * @hook wp_update_attachment_metadata
+	 *
+	 * @param array $data The new meta array.
+	 * @param int   $id   The id.
+	 *
+	 * @return array
+	 */
+	public function record_meta_update( $data, $id ) {
+		$this->media->plugin->settings->set_param( '_currrent_attachment', $id );
+		$this->media->plugin->settings->set_param( '_currrent_meta', $data );
```
Comment (pereirinha) [inline php/media/class-filter.php:749]: @DavidCramer 

What's the use case for this param `_currrent_meta`?

**Review by pereirinha — COMMENTED** [review]
@DavidCramer 

I left you a comment.

File: php/media/class-filter.php (hunk 10 above)
Comment (DavidCramer) [reply php/media/class-filter.php:749]: None yet, but doesn't hurt for a future requirement...

**Review by pereirinha — APPROVED** [review]
Great stuff.

**Pull request merged**